File: app/schemas/portfolio.py
"""
Schémas pour le portfolio.
"""
from datetime import datetime
from typing import Dict, List, Optional
from uuid import UUID

from pydantic import Field

from .base import BaseSchema, TimestampSchema, IDSchema


class PortfolioBreakdown(BaseSchema):
    """Schéma pour la répartition du portfolio."""
    
    category: str = Field(..., description="Nom de la catégorie")
    value: str = Field(..., description="Valeur dans cette catégorie")
    percentage: str = Field(..., description="Pourcentage de cette catégorie")
    count: int = Field(..., ge=0, description="Nombre d'investissements dans cette catégorie")


class TopPerformer(BaseSchema):
    """Schéma pour les meilleurs performeurs."""
    
    investment_id: UUID = Field(..., description="ID de l'investissement")
    symbol: str = Field(..., description="Symbole de l'investissement")
    name: str = Field(..., description="Nom de l'investissement")
    gain_loss_percent: str = Field(..., description="Gain/perte en pourcentage")


class PortfolioSummary(IDSchema):
    """Schéma de résumé du portfolio."""
    
    user_id: UUID = Field(..., description="ID de l'utilisateur")
    total_value: str = Field(..., description="Valeur totale du portfolio")
    total_cost: str = Field(..., description="Coût total du portfolio")
    total_gain_loss: str = Field(..., description="Gain/perte total en valeur absolue")
    total_gain_loss_percent: str = Field(..., description="Gain/perte total en pourcentage")
    diversification_score: Optional[str] = Field(None, description="Score de diversification")
    risk_score: Optional[str] = Field(None, description="Score de risque")
    currency: str = Field(..., description="Devise principale")
    investment_count: int = Field(..., ge=0, description="Nombre total d'investissements")
    performance_status: str = Field(..., description="Statut de performance")
    last_updated: datetime = Field(..., description="Dernière mise à jour")
    
    # Répartitions
    breakdown_by_country: Dict[str, PortfolioBreakdown] = Field(default_factory=dict, description="Répartition par pays")
    breakdown_by_sector: Dict[str, PortfolioBreakdown] = Field(default_factory=dict, description="Répartition par secteur")
    breakdown_by_asset_type: Dict[str, PortfolioBreakdown] = Field(default_factory=dict, description="Répartition par type d'actif")
    
    # Performeurs
    top_performers: List[TopPerformer] = Field(default_factory=list, description="Meilleurs performeurs")
    worst_performers: List[TopPerformer] = Field(default_factory=list, description="Pires performeurs")


class DiversificationAnalysis(BaseSchema):
    """Schéma pour l'analyse de diversification."""
    
    overall_score: str = Field(..., description="Score global de diversification")
    recommendations: List[Dict[str, str]] = Field(default_factory=list, description="Recommandations d'amélioration")
    detailed_analysis: Dict[str, Dict[str, str]] = Field(default_factory=dict, description="Analyse détaillée")


class PerformanceAnalysis(BaseSchema):
    """Schéma pour l'analyse de performance."""
    
    period: str = Field(..., description="Période d'analyse")
    portfolio_return: float = Field(..., description="Rendement du portfolio")
    benchmark_return: Optional[float] = Field(None, description="Rendement du benchmark")
    alpha: Optional[float] = Field(None, description="Alpha du portfolio")
    beta: Optional[float] = Field(None, description="Beta du portfolio")
    sharpe_ratio: Optional[float] = Field(None, description="Ratio de Sharpe")
    max_drawdown: Optional[float] = Field(None, description="Perte maximale")
    volatility: Optional[float] = Field(None, description="Volatilité")
    monthly_returns: List[Dict[str, float]] = Field(default_factory=list, description="Rendements mensuels")


class PortfolioSnapshot(BaseSchema):
    """Schéma pour un snapshot de portfolio."""
    
    id: UUID = Field(..., description="ID du snapshot")
    user_id: UUID = Field(..., description="ID de l'utilisateur")
    snapshot_date: str = Field(..., description="Date du snapshot")
    total_value: str = Field(..., description="Valeur totale")
    total_cost: str = Field(..., description="Coût total")
    total_gain_loss: str = Field(..., description="Gain/perte total")
    total_gain_loss_percent: str = Field(..., description="Gain/perte en pourcentage")
    diversification_score: Optional[str] = Field(None, description="Score de diversification")
    risk_score: Optional[str] = Field(None, description="Score de risque")
    currency: str = Field(..., description="Devise")
    investment_count: int = Field(..., description="Nombre d'investissements")
    created_at: datetime = Field(..., description="Date de création")


class PortfolioSnapshotList(BaseSchema):
    """Schéma pour une liste de snapshots de portfolio."""
    
    items: List[PortfolioSnapshot] = Field(..., description="Liste des snapshots")
    total: int = Field(..., description="Nombre total de snapshots")
    limit: int = Field(..., description="Limite appliquée")
    offset: int = Field(..., description="Offset appliqué")


class PortfolioComparison(BaseSchema):
    """Schéma pour la comparaison de portfolios."""
    
    current_portfolio: PortfolioSummary = Field(..., description="Portfolio actuel")
    previous_portfolio: Optional[PortfolioSummary] = Field(None, description="Portfolio précédent")
    changes: Dict[str, Dict[str, str]] = Field(default_factory=dict, description="Changements détectés")
    performance_comparison: Optional[Dict[str, float]] = Field(None, description="Comparaison de performance")

File: mobile app/lib/models/portfolio.dart
/// Safely parse a value that may be num or String to double.
double _toDouble(dynamic value, [double fallback = 0]) {
  if (value == null) return fallback;
  if (value is num) return value.toDouble();
  if (value is String) return double.tryParse(value) ?? fallback;
  return fallback;
}

int _toInt(dynamic value, [int fallback = 0]) {
  if (value == null) return fallback;
  if (value is int) return value;
  if (value is num) return value.toInt();
  if (value is String) return int.tryParse(value) ?? fallback;
  return fallback;
}

class PortfolioBreakdownItem {
  final String category;
  final double value;
  final double percentage;
  final int count;

  PortfolioBreakdownItem({
    required this.category,
    required this.value,
    required this.percentage,
    required this.count,
  });

  factory PortfolioBreakdownItem.fromJson(String key, Map<String, dynamic> json) {
    return PortfolioBreakdownItem(
      category: key,
      value: _toDouble(json['value']),
      percentage: _toDouble(json['percentage']),
      count: _toInt(json['count']),
    );
  }
}

class TopPerformer {
  final String investmentId;
  final String symbol;
  final String name;
  final double gainLossPercent;

  TopPerformer({
    required this.investmentId,
    required this.symbol,
    required this.name,
    required this.gainLossPercent,
  });

  factory TopPerformer.fromJson(Map<String, dynamic> json) {
    return TopPerformer(
      investmentId: json['investment_id'] as String,
      symbol: json['symbol'] as String,
      name: json['name'] as String,
      gainLossPercent: _toDouble(json['gain_loss_percent']),
    );
  }
}

class PortfolioMetrics {
  final String userId;
  final double totalValue;
  final double totalCost;
  final double totalGainLoss;
  final double totalGainLossPercent;
  final double diversificationScore;
  final int investmentCount;
  final List<PortfolioBreakdownItem> breakdownByCountry;
  final List<PortfolioBreakdownItem> breakdownBySector;
  final List<PortfolioBreakdownItem> breakdownByAssetType;
  final List<TopPerformer> topPerformers;
  final List<TopPerformer> worstPerformers;
  final String currency;

  PortfolioMetrics({
    required this.userId,
    required this.totalValue,
    required this.totalCost,
    required this.totalGainLoss,
    required this.totalGainLossPercent,
    required this.diversificationScore,
    required this.investmentCount,
    required this.breakdownByCountry,
    required this.breakdownBySector,
    required this.breakdownByAssetType,
    required this.topPerformers,
    required this.worstPerformers,
    required this.currency,
  });

  factory PortfolioMetrics.fromJson(Map<String, dynamic> json) {
    return PortfolioMetrics(
      userId: json['user_id'] as String,
      totalValue: _toDouble(json['total_value']),
      totalCost: _toDouble(json['total_cost']),
      totalGainLoss: _toDouble(json['total_gain_loss']),
      totalGainLossPercent: _toDouble(json['total_gain_loss_percent']),
      diversificationScore: _toDouble(json['diversification_score']),
      investmentCount: _toInt(json['investment_count']),
      breakdownByCountry: _parseBreakdown(json['breakdown_by_country']),
      breakdownBySector: _parseBreakdown(json['breakdown_by_sector']),
      breakdownByAssetType: _parseBreakdown(json['breakdown_by_asset_type']),
      topPerformers: (json['top_performers'] as List?)
              ?.map((e) => TopPerformer.fromJson(e as Map<String, dynamic>))
              .toList() ??
          [],
      worstPerformers: (json['worst_performers'] as List?)
              ?.map((e) => TopPerformer.fromJson(e as Map<String, dynamic>))
              .toList() ??
          [],
      currency: json['currency'] as String? ?? 'USD',
    );
  }

  static List<PortfolioBreakdownItem> _parseBreakdown(dynamic data) {
    if (data == null || data is! Map<String, dynamic>) return [];
    return data.entries
        .map((e) => PortfolioBreakdownItem.fromJson(e.key, e.value as Map<String, dynamic>))
        .toList();
  }
}

class PortfolioHistoryPoint {
  final String timestamp;
  final double totalValue;
  final double totalCost;
  final double totalGainLoss;

  PortfolioHistoryPoint({
    required this.timestamp,
    required this.totalValue,
    required this.totalCost,
    required this.totalGainLoss,
  });

  DateTime get date => DateTime.parse(timestamp);

  factory PortfolioHistoryPoint.fromJson(Map<String, dynamic> json) {
    return PortfolioHistoryPoint(
      timestamp: json['timestamp'] as String,
      totalValue: (json['total_value'] as num?)?.toDouble() ?? 0,
      totalCost: (json['total_cost'] as num?)?.toDouble() ?? 0,
      totalGainLoss: (json['total_gain_loss'] as num?)?.toDouble() ?? 0,
    );
  }
}

class PortfolioHistoryResponse {
  final String userId;
  final List<PortfolioHistoryPoint> dataPoints;
  final int totalPoints;
  final String? startDate;
  final String? endDate;

  PortfolioHistoryResponse({
    required this.userId,
    required this.dataPoints,
    required this.totalPoints,
    this.startDate,
    this.endDate,
  });

  factory PortfolioHistoryResponse.fromJson(Map<String, dynamic> json) {
    return PortfolioHistoryResponse(
      userId: json['user_id'] as String,
      dataPoints: (json['data_points'] as List)
          .map((e) => PortfolioHistoryPoint.fromJson(e as Map<String, dynamic>))
          .toList(),
      totalPoints: json['total_points'] as int,
      startDate: json['start_date'] as String?,
      endDate: json['end_date'] as String?,
    );
  }
}
